core/trino-main/src/main/java/io/trino/operator/window/matcher/MatchResult.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.window.matcher;
+
+import static java.util.Objects.requireNonNull;
+
+public class MatchResult
+{
+    public static final MatchResult NO_MATCH = new MatchResult(false, new int[] {}, new int[] {});
+
+    private final boolean matched;
+    private final int[] labels;
+    private final int[] exclusions;
+
+    public MatchResult(boolean matched, int[] labels, int[] exclusions)
+    {
+        this.matched = matched;
+        this.labels = requireNonNull(labels, "labels is null");
+        this.exclusions = requireNonNull(exclusions, "exclusions is null");
```

Comment:
The array that's passed in is being assigned directly and then returned via the getLabels/getExclusions methods. That's potentially unsafe, depending on what the caller of this constructor is doing with the arrays after. We should either 
1) copy the arrays
2) use a proper wrapper that guarantees immutable semantics (both from the caller of MatchResult modifying the contents after the constructor exits and from the caller of getLabels/getExclusions from modifying the result)

Reply:
I wrapped the returned arrays of labels and exclusions in an object that doesn't allow modifications. It only allows to access array elements and array length.
This solves the issue of copying the array as well as the issue of modifying the array by the caller.

However, this is not secured against the underlying tables being modified by the owner (`Captures`)